lib/jekyll/collection.rb

```diff
@@ -210,6 +210,29 @@ def read_document(full_path)
 
     private
 
+    def sort_docs!
+      if @metadata["order"] && @metadata["order"].is_a?(Array)
+        rearrange_docs!
+      elsif @metadata["sort_by"]
+        docs.sort_by! { |d| d.data[@metadata["sort_by"]] }
+      else
+        docs.sort!
+      end
+    end
+
+    def rearrange_docs!
+      @metadata["order"].each do |o|
+        doc_path = "_#{@label}/#{o}"
+        doc = docs.find { |d| d.relative_path == doc_path }
+        if File.exist?(@site.in_source_dir(doc_path))
+          doc.data["sort_index"] = @metadata["order"].index(o)
+        else
+          Jekyll.logger.warn "Error:", "#{o} not found in '#{@label}' collection!"
+        end
+      end
+      docs.sort_by! { |d| d.data["sort_index"] }
+    end
```

Comment:
1.  What happens to files that are in the collection directory but *not* specified in the explicit order list in config?
2.  What happens to files that are in the collection directory but do not specify the key being used for sorting (in this case, don't have a `lesson` key)?